CIP-plutus-modules/README.md

```diff
@@ -0,0 +1,1622 @@
+transaction, either directly or as reference scripts.
+
+The advantage of this approach is that only the code in the balancer
+needs to be changed. The disadvantage is that transaction balancing
+becomes more expensive: script verification may need to be rerun up to
+once per script or reference script. In comparison to the first
+approach above, this one is more complex to implement, but replaces a
+quadratic algorithm by a linear one.
+
+##### Third approach: modified CEK machine
+
+The most direct way to determine that a script is not redundant is to
+observe it being executed during script verification. Unfortunately,
+the CEK machine, in its present form, does not make that
+possible. Thus an alternative is to *modify the CEK machine* so that a
+balancer can observe scripts being executed, and declare all the other
+scripts redundant. In comparison to the first two approaches, this is
+likely to be much more efficient, because it only requires running
+script verification once.
+
+The modifications needed to the CEK machine are as follows:
+
+`CekValue`s are extended with *tagged values*, whose use can be
+observed in the result of a run of the machine.
+```
+data CekValue uni fun ann =
+  ...
+  | VTag ScriptHash (CekValue uni fun ann)
+```
+In the 'value script' variation, no expression resulting in a `VTag`
+value is needed, because `VTag`s will be inserted only by
+`resolveScriptDependencies`. In other variations, a `Tag` constructor
+must also be added to the `NTerm` type, to be added by
+`resolveScriptDependencies`. In either case the version of
+`runScriptDependencies` *used in the balancer* tags each value or
+subterm derived from a `ScriptHash` `h` as `VTag h ...` (or `Tag h
+...` in variations other than 'value scripts').
+
+The CEK machine is parameterized over an emitter function, used for
+logging. We can make use of this to emit `ScriptHash`es as they are
+used. This allows the balancer to observe which `ScriptHash`es *were*
+used.
+
+Simply evaluating a tagged value, or building it into a
+data-structure, does not *use* it in the sense we mean here: replacing
+such a value with `builtin unit` will not cause a validation
+failure. Only when such a value is actually *used* should we strip the
+tag, emit the `ScriptHash` in the `CekM` monad, and continue with the
+untagged value. This should be done in `returnCek`, on encountering a
+`FrameCases` context for a tagged value, and in `applyEvaluate` when
+the function to be applied turns out to be tagged, or when the
+argument to a `builtin` turns out to be tagged.
+
+Adding and removing tags must be assigned a zero cost *in the
+balancer*, since the intention is that they should not appear in
+transactions when they are verified on the chain. Thus a zero cost is
+required for the balancer to return accurate costs for script
+verification on the chain. On the other hand, if these operations *do*
+reach the chain, then they should have a *high* cost, to deter attacks

```

Comment:
they should just not be part of the specified syntax of UPLC and be invalid on chain, easy

Reply:
> > If modules are allowed to be arbitrary terms then you could supply let _ = error in () and the script would fail, but if that's replaced with () then the script might suceed.

> This is a good point! I hadn't though that this might convert failures to successes but indeed it can. This doesn't occur with the value scripts proposal, which is maybe a point in its favour.

No. In this case the *original* attempt to verify the transaction, with *all* modules present, will fail. The point of lazy loading is only to preserve success--there's no need even to try dropping reference inputs from transactions which already fail.

> > I'm not sure that the script would necessarily produce a run-time type error if it tries to use a supposedly unused module which has been replaced by builtin unit.

> Well, I think the claim is that either it produces the same result as it would have if we hadn't replaced it or it gives an error. And I think that's pretty convincing? All you can do with a unit value is:

> Do things which don't depend on what it is at all (semantics are the same)
> Do things that rely on it being unit (semantics are the same!)
> Do things that rely on it being non-unit (should fail when given unit)

In *untyped* code, you can check whether you have unit (I assume), and do something else if you do. This would enable untyped scripts to detect that a module was missing, and handle that error somehow. Not sure if that's useful, but it seems legitimate at least. UPLC compiled from typed code will not be able to do this, of course.

Reply:
> they should just not be part of the specified syntax of UPLC and be invalid on chain, easy

You mean, include them in the type, but don't deserialize them? Seems a little brittle: a future developer might notice the missing case in the deserializer, and add it, not realising that its absence was *essential* to defend against an attack. Or am I just jaded?

Reply:
> No. In this case the original attempt to verify the transaction, with all modules present, will fail. The point of lazy loading is only to preserve success--there's no need even to try dropping reference inputs from transactions which already fail.

You're not thinking about attackers. If the version with the modules present fails but the version with some moduels present doesn't, then that's potentially an _attack_. We really do not want to let the person running the script make it succeed when it would have failed! 

> You mean, include them in the type, but don't deserialize them? Seems a little brittle: a future developer might notice the missing case in the deserializer, and add it, not realising that its absence was essential to defend against an attack.

We have a specification for a reason :)

Reply:
> > No. In this case the original attempt to verify the transaction, with all modules present, will fail. The point of lazy loading is only to preserve success--there's no need even to try dropping reference inputs from transactions which already fail.
> 
> You're not thinking about attackers. If the version with the modules present fails but the version with some moduels present doesn't, then that's potentially an _attack_. We really do not want to let the person running the script make it succeed when it would have failed!

Hmm. We're considering a case in which a module fails when evaluated, causing any script which imports it to fail. Note that this can't happen with value scripts. But in that case the vulnerability would kick in if somebody put a script on the chain that DOES import the module in question, and so can never succeed, and then an attacker USED that script, but left out the offending module. Suppose the script is a spending verifier on a UTxO. Then the vulnerable situation is where someone creates a UTxO with a spending verifier that can never succeed, and does so in this slightly obscure manner, but an attacker can spend the UTxO anyway. OK, it's a vulnerability. You really have to work to fall victim to it though! And it is interesting that the value scripts idea fixes it...

Reply:
You have to work to fall victim to it... if you use this feature in the way you are envisaging. Which people won't. IME it's just better not to have such semantic loopholes at all, if we can possibly avoid it.

Reply:
Well, that's an argument in favour of the value scripts variation.